src/Avalonia.X11/X11TrayIconImpl.cs

```diff
@@ -0,0 +1,374 @@
+            _isActive = false;
+        }
+
+        public void Dispose()
+        {
+            _isDisposed = true;
+            DestroyTrayIcon();
+            con.Dispose();
+        }
+
+        public void SetIcon(IWindowIconImpl icon)
+        {
+            if (con == null || _isDisposed) return;
+            if (!(icon is X11IconData x11icon)) return;
+
+            var w = (int)x11icon.Data[0];
+            var h = (int)x11icon.Data[1];
+
+            var rx = x11icon.Data.AsSpan(2);
+            var pixLength = w * h;
+
+            var pixByteArrayCounter = 0;
+            var pixByteArray = new byte[w * h * 4];
+
+            for (var i = 0; i < pixLength; i++)
+            {
+                var u = rx[i].ToUInt32();
+                pixByteArray[pixByteArrayCounter++] = (byte)((u & 0xFF000000) >> 24);
+                pixByteArray[pixByteArrayCounter++] = (byte)((u & 0xFF0000) >> 16);
+                pixByteArray[pixByteArrayCounter++] = (byte)((u & 0xFF00) >> 8);
+                pixByteArray[pixByteArrayCounter++] = (byte)(u & 0xFF);
+            }
+
+            _icon = new DbusPixmap(w, h, pixByteArray);
+            _statusNotifierItemDbusObj.SetIcon(_icon);
+        }
+        
+        public void SetIsVisible(bool visible)
+        {
+            if (con == null || _isDisposed || !_ctorFinished) return;
+
+            if (visible & !_isActive)
+            {
+                DestroyTrayIcon();
+                CreateTrayIcon();
+            }
+            else if (!visible & _isActive)
+            {
+                DestroyTrayIcon();
+            }
+        }
+
+        public void SetToolTipText(string text)
+        {
+            if (con == null || _isDisposed) return;
+            _tooltipText = text;
+            _statusNotifierItemDbusObj?.SetTitleAndTooltip(_tooltipText);
+        }
+    }
+    
+    /// <summary>
+    /// DBus Object used for setting system tray icons.
+    /// </summary>
+    /// <remarks>
+    /// Useful guide: https://web.archive.org/web/20210818173850/https://www.notmart.org/misc/statusnotifieritem/statusnotifieritem.html
+    /// </remarks>
+    internal class StatusNotifierItemDbusObj : IStatusNotifierItem
+    {
+        private readonly StatusNotifierItemProperties _backingProperties;
+        private event Action<PropertyChanges> OnPropertyChange;
+        public event Action OnTitleChanged;
+        public event Action OnIconChanged;
+        public event Action OnAttentionIconChanged;
+        public event Action OnOverlayIconChanged;
+        public event Action OnTooltipChanged;
+        public Action<INativeMenuExporter> SetNativeMenuExporter { get; set; }
+        public Action<string> NewStatusAsync { get; set; }
+        public Action ActivationDelegate { get; set; }
+        public ObjectPath ObjectPath { get; }
+
+        public StatusNotifierItemDbusObj(ObjectPath dbusmenuPath)
+        {
+            ObjectPath = new ObjectPath($"/StatusNotifierItem");
```

Comment:
Do we always use the same hardcoded object path? What about multiple tray icons in the same app?

Reply:
yes... sadly this is what the spec says we should do. on the case of multiple icons, the separate dbus session connections should not conflict with each other.

Reply:
any source?

Reply:
https://www.freedesktop.org/wiki/Specifications/StatusNotifierItem/

Reply:
it was agreed that the spec is broken and that current approach is good.

discussion was with @danwalmsley @kekekeks and @jmacato